File: backend/app/schemas/reverse_proxy.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ReverseProxyRuleCreate(BaseModel):
    vmid: int = Field(gt=0)
    zone_id: str = Field(min_length=1, max_length=64)
    hostname_prefix: str = Field(default="", max_length=190)
    internal_port: int = Field(ge=1, le=65535)
    enable_https: bool = True


class ReverseProxyRuleUpdate(ReverseProxyRuleCreate):
    pass


class ReverseProxyZoneOption(BaseModel):
    id: str
    name: str


class ReverseProxySetupContext(BaseModel):
    enabled: bool
    gateway_ready: bool
    cloudflare_ready: bool
    reasons: list[str] = Field(default_factory=list)
    zones: list[ReverseProxyZoneOption] = Field(default_factory=list)
    default_dns_target_type: str | None = None
    default_dns_target_value: str | None = None


class ReverseProxyRuntimeSection(BaseModel):
    routers: list[dict[str, Any]] = Field(default_factory=list)
    services: list[dict[str, Any]] = Field(default_factory=list)
    middlewares: list[dict[str, Any]] = Field(default_factory=list)


class ReverseProxyRuntimeSnapshot(BaseModel):
    runtime_error: str | None = None
    version: dict[str, Any] | None = None
    overview: dict[str, Any] | None = None
    entrypoints: list[dict[str, Any]] = Field(default_factory=list)
    http: ReverseProxyRuntimeSection = Field(default_factory=ReverseProxyRuntimeSection)
    tcp: ReverseProxyRuntimeSection = Field(default_factory=ReverseProxyRuntimeSection)
    udp: ReverseProxyRuntimeSection = Field(default_factory=ReverseProxyRuntimeSection)


__all__ = [
    "ReverseProxyRuleCreate",
    "ReverseProxyRuleUpdate",
    "ReverseProxyZoneOption",
    "ReverseProxySetupContext",
    "ReverseProxyRuntimeSection",
    "ReverseProxyRuntimeSnapshot",
]

File: frontend/src/services/reverseProxy.ts
import type { CancelablePromise } from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

export type TraefikRuntimeItem = Record<string, unknown>

export type TraefikRuntimeSection = {
  routers: TraefikRuntimeItem[]
  services: TraefikRuntimeItem[]
  middlewares: TraefikRuntimeItem[]
}

export type TraefikRuntimeSnapshot = {
  runtime_error?: string | null
  version?: Record<string, unknown> | null
  overview?: Record<string, unknown> | null
  entrypoints: TraefikRuntimeItem[]
  http: TraefikRuntimeSection
  tcp: TraefikRuntimeSection
  udp: TraefikRuntimeSection
}

export type ManagedReverseProxyRule = {
  id: string
  vmid: number
  vm_ip: string
  domain: string
  zone_id?: string | null
  internal_port: number
  enable_https: boolean
  dns_provider: string
  created_at: string
}

export type ReverseProxyRuleCreateInput = {
  vmid: number
  zone_id: string
  hostname_prefix: string
  internal_port: number
  enable_https: boolean
}

export type ReverseProxyZoneOption = {
  id: string
  name: string
}

export type ReverseProxySetupContext = {
  enabled: boolean
  gateway_ready: boolean
  cloudflare_ready: boolean
  reasons: string[]
  zones: ReverseProxyZoneOption[]
  default_dns_target_type?: string | null
  default_dns_target_value?: string | null
}

export const ReverseProxyApiService = {
  getRuntimeSnapshot(): CancelablePromise<TraefikRuntimeSnapshot> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/reverse-proxy/runtime",
    })
  },

  listRules(): CancelablePromise<ManagedReverseProxyRule[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/reverse-proxy/rules",
    })
  },

  getSetupContext(): CancelablePromise<ReverseProxySetupContext> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/reverse-proxy/setup-context",
    })
  },

  createRule(
    requestBody: ReverseProxyRuleCreateInput,
  ): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/reverse-proxy/rules",
      body: requestBody,
      mediaType: "application/json",
    })
  },

  updateRule(data: {
    ruleId: string
    requestBody: ReverseProxyRuleCreateInput
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "PUT",
      url: "/api/v1/reverse-proxy/rules/{rule_id}",
      path: { rule_id: data.ruleId },
      body: data.requestBody,
      mediaType: "application/json",
    })
  },

  deleteRule(data: { ruleId: string }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/reverse-proxy/rules/{rule_id}",
      path: { rule_id: data.ruleId },
    })
  },

  syncRules(): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/reverse-proxy/rules/sync",
    })
  },
}
